components/cdc/src/observer.rs

```diff
@@ -164,29 +202,137 @@ impl RegionChangeObserver for CdcObserver {
+        let user_key = Key::truncate_ts_for(key.as_encoded()).unwrap();
+        if self
+            .write_cursor
+            .near_seek(key, &mut self.statistics.write)?
+            && Key::is_user_key_eq(self.write_cursor.key(&mut self.statistics.write), user_key)
+        {
+            if self.write_cursor.key(&mut self.statistics.write) == key.as_encoded().as_slice() {
+                // Key was committed, move cursor to the next key to seek for old value.
+                if !self.write_cursor.next(&mut self.statistics.write) {
+                    // Do not has any next key, return empty value.
+                    return Ok(Some(Vec::default()));
+                }
+            } else if !self.check_lock(key) {
+                return Ok(None);
+            }
+
+            // Key was not committed, check if the lock is corresponding to the key.
+            let mut old_value = Some(Vec::default());
+            while Key::is_user_key_eq(self.write_cursor.key(&mut self.statistics.write), user_key) {
+                let write =
+                    WriteRef::parse(self.write_cursor.value(&mut self.statistics.write)).unwrap();
+                old_value = match write.write_type {
+                    WriteType::Put => match write.short_value {
+                        Some(short_value) => Some(short_value.to_vec()),
+                        None => {
+                            let key = key.clone().truncate_ts().unwrap().append_ts(write.start_ts);
+                            self.get_value_default(&key)
+                        }
+                    },
+                    WriteType::Delete => Some(Vec::default()),
+                    WriteType::Rollback | WriteType::Lock => {
+                        if !self.write_cursor.next(&mut self.statistics.write) {
+                            Some(Vec::default())
+                        } else {
+                            continue;
+                        }
+                    }
+                };
+                break;
+            }
+            Ok(old_value)
+        } else if self.check_lock(key) {
+            Ok(Some(Vec::default()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
+    use engine_rocks::RocksEngine;
     use kvproto::metapb::Region;
     use kvproto::raft_cmdpb::*;
     use std::time::Duration;
+    use tikv::storage::kv::TestEngineBuilder;
+    use tikv::storage::mvcc::tests::*;
 
     #[test]
     fn test_register_and_deregister() {
         let (scheduler, rx) = tikv_util::worker::dummy_scheduler();
         let observer = CdcObserver::new(scheduler);
         let observe_id = ObserveID::new();
+        let engine = TestEngineBuilder::new().build().unwrap().get_rocksdb();
 
-        observer.on_prepare_for_apply(observe_id, 0);
-        observer.on_apply_cmd(
+        <CdcObserver as CmdObserver<RocksEngine>>::on_prepare_for_apply(&observer, observe_id, 0);
```

Comment:
It looks bizarre. 😅

Reply:
Yes, it is for specifying the engine parameter.😢